plugins/corsair/fu-corsair-bp.c

```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright (C) 2022 Andrii Dushko <andrii.dushko@developex.net>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-corsair-bp.h"
+#include "fu-corsair-common.h"
+
+#define CORSAIR_DEFAULT_VENDOR_INTERFACE_ID 1
+#define CORSAIR_ACTIVATION_TIMEOUT	    30000
+#define CORSAIR_MODE_BOOTLOADER		    3
+#define CORSAIR_FIRST_CHUNK_HEADER_SIZE	    7
+#define CORSAIR_NEXT_CHUNKS_HEADER_SIZE	    3
+#define CORSAIR_TRANSACTION_TIMEOUT	    10000
+#define CORSAIR_DEFAULT_CMD_SIZE	    64
+
+#define CORSAIR_SUBDEVICE_MASK 0x01
+
+#define CORSAIR_OFFSET_CMD_PROPERTY_ID	  0x02
+#define CORSAIR_OFFSET_CMD_PROPERTY_VALUE 0x03
+#define CORSAIR_OFFSET_CMD_VERSION	  0x03
+#define CORSAIR_OFFSET_CMD_CRC		  0x08
+#define CORSAIR_OFFSET_CMD_MODE		  0x03
+#define CORSAIR_OFFSET_CMD_STATUS	  0x02
+#define CORSAIR_OFFSET_CMD_FIRMWARE_SIZE  0x03
+#define CORSAIR_OFFSET_CMD_SET_MODE	  0x04
+#define CORSAIR_OFFSET_CMD_DESTINATION	  0x00
+
+struct _FuCorsairBp {
+	GObject parent_instance;
+	GUsbDevice *usb_device;
```

Comment:
This is a bit fishy; why not derive from FuUsbDevice?

Reply:
My intention was to move low-level protocol stuff out of FuCorsairDevice that's why I passed GUsbDevice. I think it may be confusing that it derives from a device and don't behaves like a device.

Reply:
Would it be clearer if the BP device was a separate device child? e.g. 

    ├─MX Master 3 B:
    │ │   Device ID:        f3d644e5bf195ada87526fbea0e8af23386c3e71
    │ │   Summary:          Unifying Mouse
    │ │   Current version:  RBM14.00_B0008
    │ │   Bootloader Version:BL141.00_B0008
    │ │   Vendor:           Logitech (USB:0x046D)
    │ │   Install Duration: 4 minutes
    │ │   Update Error:     Battery level is too low
    │ │   GUID:             959e9b9b-2071-5d13-8e7e-c8a76795a68c ← HIDRAW\VEN_046D&MOD_B02800000000
    │ │   Device Flags:     • Updatable
    │ │                     • Signed Payload
    │ │ 
    │ └─Radio:
    │       Device ID:      cc9a4150c989378431a2ae92e57fec98f9e31a4f
    │       Current version:0x0119
    │       Vendor:         Logitech (USB:0x046D)
    │       Install Duration:4 minutes
    │       GUID:           8a8b36a9-9fc1-5d70-ab7f-c6563e4955cd ← HIDRAW\VEN_046D&MOD_B02800000000&ENT_05
    │       Device Flags:   • Updatable

Reply:
yes, it shows like this:

> ├─KATAR PRO WIRELESS receiver:
│ │   Device ID:          0d5d05911800242bb1f35287012cdcbd9b381148
│ │   Current version:    2.3.21
│ │   Bootloader Version: 1.0.8
│ │   Vendor:             USB:0x1B1C
│ │   GUIDs:              8bf1f03f-3c6f-53a1-9169-800fddf0952c ← USB\VID_1B1C&PID_1B94
│ │                       d887d9d9-09c6-5852-9d27-c99b36c627bb ← USB\VID_1B1C&PID_1B94&REV_0203
│ │   Device Flags:       • Updatable
│ │                       • Unsigned Payload
│ │ 
│ └─KATAR PRO WIRELESS mouse:
│       Device ID:        02b076547db3e4f6e3993c59b6824f7670ece79c
│       Current version:  2.5.30
│       Bootloader Version:2.0.9
│       Vendor:           USB:0x1B1C
│       GUID:             7235bc76-2e5a-5312-b848-882871503152 ← USB\VID_1B1C&PID_1B94&WIRELESS
│       Device Flags:     • Updatable
│                         • Unsigned Payload

BP is not a device, but protocol routines which device (devices in future) will use. I will add readme and more comments.

Reply:
> BP is not a device, but protocol routines

That's what we would call a "_proxy_" -- i.e. you can create a FuCorsairBp object and add it to the parent device with fu_device_set_proxy() -- then whenever you need the BP device you can do fu_device_get_proxy(). Using the proxy functionality means all the refcounting gets done automatically for you. The proxy doesn't get "added" to the daemon as a visible object, it's just used in the device subclass.

Reply:
I'll try to use proxy functionality. thank you

Reply:
As far as I understood I still need to do refcounting by myself because fu_device_set_proxy() uses weak ref. That's why I think I wouldn't get any benefits from using proxy over self->bp because each FuCorsairDevice has its own FuCorsairBp instance.

Reply:
That's a good point, agreed.